mautic/mautic
# Fix #10877. Migration Version20201120122846: Improve logic to check table and its foreign keys

Fix #10877: The logic to check if the table exists and if it has foreign keys is weak and causes issues when foreign keys have different names (as it frequently happens).



| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | 4.1
| New feature/enhancement? (use the a.x branch)      | No
| Deprecations?                          | [No
| BC breaks? (use the c.x branch)        | [No
| Automated tests included?              | No
| Related user documentation PR URL      |
| Related developer documentation PR URL | 
| Issue(s) addressed                     | Fixes #10877, #10870, #10865, #9129



#### Description:

The problem comes from the foreign keys: the migration assumes the names of the foreign keys are deterministic, but they aren't, and this causes the issue.

The check, in fact, groups the creation of the table and the creation of the foreign keys in one condition, but this is wrong, as maybe the table is there and the names of foreign keys are different than the ones generated by the migration: if names are different (and they are different), then the migration fail.

The refactored logic is now more solid and avoids issues running the migration, applying only the queries really necessary.

More details here: https://github.com/mautic/mautic/issues/9129#issuecomment-1044958793

#### Steps to test this PR:

This is not always testable: it works only if you updated the database schema using `bin/console doctrine:schema:update --force`.

This command, in fact, creates the foreign keys with "arbitrary" names: in a case like this, the new logic can handle anyway the checks to verify if foreign keys already exist and, if they exist, skip the query to create them.

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2.


<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/10878"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
